File: flutter_alpha_player_plugin/ios/Classes/NativeAlphaPlayer.swift
import Flutter
import UIKit

class NativeAlphaPlayer: NSObject, FlutterPlatformView, PlayerViewDelegate {
    /// 通信通道
    private var methodChannel: FlutterMethodChannel?
    /// 传进来的坐标
    private var frame: CGRect?
    /// flutter 传的参数
    private var arguments: Any?
    
    /// 原生视图view
    lazy var playerView: PlayerView = {
        let v = PlayerView()
        v.delegate = self
        return v
    }()

    /// 自定义初始化方法
    init(frame: CGRect, viewIdentifier: Int64, arguments: Any, messenger: FlutterBinaryMessenger) {
        super.init()
        self.frame = frame
        self.arguments = arguments
        /// 建立通信通道 用来 监听Flutter 的调用和 调用Fluttter 方法 这里的名称要和Flutter 端保持一致
        methodChannel = FlutterMethodChannel(name: "flutter_alpha_player_plugin_\(viewIdentifier)", binaryMessenger: messenger)
        methodChannel?.setMethodCallHandler {
            [weak self] (call: FlutterMethodCall, result: FlutterResult) in
            // 注意闭包引用 内存无法释放
            self?.handleMethod(call: call, result: result)
        }
    }

    func view() -> UIView {
        playerView.frame = frame!
        playerView.clipsToBounds = true
        return playerView
    }

    // MARK: - flutter 调 ios 回调

    /// 接收flutter发来的消息
    func handleMethod(call: FlutterMethodCall, result: FlutterResult) {
        switch call.method {
        /// 开始播放
        case "play":
            guard let params = call.arguments as? [String: Any],
                  let path = params["filePath"] as? String,
                  !path.isEmpty else {
                errorAction(code: 1001, message: "filePath is empty or null.")
                result(0)
                return
            }

            playerView.addAlphaPlayerViewToParentView()
            let scaleType = params["scaleType"] as? Int
            playerView.play(path: path, scaleType: scaleType)
            result(0)
            break
        /// 停止播放
        case "stop": fallthrough
        case "release":
            playerView.playerStopWithFinishPlayingCallback()
            playerView.removeAlphaPlayerViewFromSuperView()
            result(0)
            break
        /// 同步视图
        case "attachView":
            playerView.addAlphaPlayerViewToParentView()
            result(0)
            break
        /// 移除视图
        case "detachView":
            playerView.removeAlphaPlayerViewFromSuperView()
            result(0)
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }

    // MARK: - ios ---> flutter 事件回调

    // @param method 方法名称，唯一关系绑定，
    // @param arguments 参数或者数据
    private func callFlutterMethod(method: String, arguments: Any?) {
        methodChannel?.invokeMethod(method, arguments: arguments)
    }

    // MARK: - PlayerViewDelegate

    /// 开始播放
    func alphaPlayerStartPlay() {
        callFlutterMethod(method: "play", arguments: nil)
    }

    /// 播放结束回调 isNormalFinsh 是否正常播放结束 （True 是  false 播放报错）
    func alphaPlayerDidFinishPlaying(isNormalFinsh: Bool, errorStr: String?) {
        callFlutterMethod(method: "stop", arguments: nil)
    }

    /// 回调每一帧的持续时间
    func videoFrameCallBack(duration: TimeInterval) {
    }

    /// 错误回调
    func errorAction(code: Int, message: String) {
        callFlutterMethod(method: "error", arguments: [
            "code": code,
            "message": message,
        ] as [String : Any])
    }

    deinit {
        print("--- NativeAlphaPlayer deinit ---")
    }
}

File: flutter_alpha_player_plugin/android/src/main/kotlin/com/lyg/flutter_alpha_player_plugin/NativeAlphaPlayer.kt
package com.lyg.flutter_alpha_player_plugin

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.view.View
import com.ss.ugc.android.alpha_player.IMonitor
import com.ss.ugc.android.alpha_player.IPlayerAction
import com.ss.ugc.android.alpha_player.model.ScaleType
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

/// 原生界面需要实现PlatformView
internal class NativeAlphaPlayer(
        binaryMessenger: BinaryMessenger,
        context: Context?,
        id: Int?,
        createParams: Map<String, Any>?) : MethodChannel.MethodCallHandler, PlatformView {

    private val mContext: Context? = context
    private lateinit var playerView: VideoView
    private lateinit var channel: MethodChannel
    private var handler: Handler = Handler(Looper.getMainLooper())


    init {
        mContext?.let {
            playerView = VideoView(it)
            playerView.initPlayerController(it, object : IPlayerAction {
                override fun endAction() {
                    // 主动释放内存
                    playerView.reset()
                    callFlutterMethod("stop", null)
                }

                override fun onVideoSizeChanged(videoWidth: Int, videoHeight: Int, scaleType: ScaleType) {
                    print("videoWidth: $videoWidth videoHeight: $videoHeight scaleType: $scaleType")
                }

                override fun startAction() {
                    callFlutterMethod("play", null)
                }

                override fun errorAction(code: Int, message: String) {
                    val map = hashMapOf(
                            "code" to code,
                            "message" to message
                    )
                    callFlutterMethod("error", map)
                }
            }, object : IMonitor {
                override fun monitor(
                        result: Boolean,
                        playType: String,
                        what: Int,
                        extra: Int,
                        errorInfo: String
                ) {
                    print("result: $result playType: $playType what: $what extra: $extra errorInfo: $errorInfo")
                }
            }
            )

            playerView.attachView()
            channel = MethodChannel(binaryMessenger, "flutter_alpha_player_plugin_${id}")
            channel.setMethodCallHandler(this);
        }
    }

    // MethodChannel 回调 Flutter-》Android
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            // 播放视频
            "play" -> {
                val path = call.argument<String>("filePath")
                val scaleType = call.argument<Int?>("scaleType") ?: 2
                playerView.start(path, scaleType, scaleType, false)
                result.success(0)
            }
            // 停止
            "stop" -> {
                playerView.stop()
                // 主动释放内存
                playerView.reset()
                result.success(0)
            }
            // 释放
            "release" -> {
                playerView.releasePlayerController()
                result.success(0)
            }
            // 同步视图
            "attachView" -> {
                playerView.attachView()
                result.success(0)
            }
            // 移除视图
            "detachView" -> {
                playerView.detachView()
                result.success(0)
            }
        }
    }
    // android -> flutter 
    /**
     * @param method 方法名称，唯一关系绑定，
     * @param arguments 参数或者数据 目前默认json
     */
    private fun callFlutterMethod(method: String, arguments: Any?) {
        handler.post {
            channel.invokeMethod(method, arguments)
        }
    }

    // 获取View
    override fun getView(): View? {
        return playerView;
    }

    // 销毁View
    override fun dispose() {
        playerView.detachView()
        playerView.releasePlayerController()
        println("--- NativeAlphaPlayer dispose ---")
    }
}